dashboard/app/controllers/api_controller.rb

```diff
@@ -175,14 +175,14 @@ def section_progress
     script = load_script(section)
 
     # stage data
-    stages = script.script_levels.where(bonus: nil).group_by(&:stage).map do |stage, levels|
+    stages = script.script_levels.select{|sl| sl.bonus.nil?}.group_by(&:stage).map do |stage, levels|
       {
         length: levels.length,
         title: ActionController::Base.helpers.strip_tags(stage.localized_title)
       }
     end
 
-    script_levels = script.script_levels.where(bonus: nil)
+    script_levels = script.script_levels.select {|sl| sl.bonus.nil?}
```

Comment:
This is better because bonus is not in our index presumably?

Reply:
no, this is unrelated to table indexes, it has to do with how ActiveRecord associations work with `#where`. Any `where` clause will construct a new SQL query, re-load data from the DB and build new ActiveRecord objects in memory, which is sub-optimal when an association is already loaded into memory (which is the case for our pre-loaded cached scripts). Instead of using `where`, it's usually better to iterate on existing records already loaded in memory via `to_a` (or filtered via `#select`, or a single record fetched via `#detect`/`#find`).

This pattern pops up quite often in our codebase actually (see e.g., #18532). It's theoretically possible to patch Rails to detect and convert a `where` query on a pre-loaded association into a corresponding in-memory fetch and I've looked into this a bit, but it's non-trivial for it handle all the various edge-cases that might present.

Reply:
Good to know. I'll try to watch out for this in future code I write/code reviews.